python/ray/serve/tests/test_backend_state.py

```diff
@@ -616,32 +636,298 @@ def test_deploy_new_config_new_version(mock_backend_state):
     assert goal_manager.check_complete(update_goal)
 
 
+def test_initial_deploy_no_throttling(mock_backend_state):
+    # All replicas should be started at once for a new deployment.
+    backend_state, timer, goal_manager = mock_backend_state
+
+    assert len(backend_state._replicas) == 0
+
+    tag = "tag"
+    b_config_1, r_config_1 = generate_configs(10)
+    goal_1 = backend_state.deploy_backend(
+        tag, b_config_1, r_config_1, version="1")
+
+    backend_state.update()
+    assert backend_state._replicas[tag].count() == 10
+    assert backend_state._replicas[tag].count(
+        version="1", states=[ReplicaState.STARTING]) == 10
+    assert not goal_manager.check_complete(goal_1)
+
+    for replica in backend_state._replicas[tag].get():
+        replica._actor.set_ready()
+
+    backend_state.update()
+
+    # Check that the new replicas have started.
+    backend_state.update()
+    assert backend_state._replicas[tag].count(
+        version="1", states=[ReplicaState.RUNNING]) == 10
+    assert goal_manager.check_complete(goal_1)
+
+
+def test_new_version_deploy_throttling(mock_backend_state):
+    # All replicas should be started at once for a new deployment.
+    # When the version is updated, it should be throttled.
+    backend_state, timer, goal_manager = mock_backend_state
+
+    assert len(backend_state._replicas) == 0
+
+    tag = "tag"
+    b_config_1, r_config_1 = generate_configs(10)
+    goal_1 = backend_state.deploy_backend(
+        tag, b_config_1, r_config_1, version="1")
+
+    backend_state.update()
+    assert backend_state._replicas[tag].count() == 10
+    assert backend_state._replicas[tag].count(
+        version="1", states=[ReplicaState.STARTING]) == 10
+    assert not goal_manager.check_complete(goal_1)
+
+    for replica in backend_state._replicas[tag].get():
+        replica._actor.set_ready()
+
+    backend_state.update()
+
+    # Check that the new replicas have started.
+    backend_state.update()
+    assert backend_state._replicas[tag].count(
+        version="1", states=[ReplicaState.RUNNING]) == 10
+    assert goal_manager.check_complete(goal_1)
+
+    # Now deploy a new version. Two old replicas should be stopped.
+    goal_2 = backend_state.deploy_backend(
+        tag, b_config_1, r_config_1, version="2")
+    backend_state.update()
+    assert backend_state._replicas[tag].count() == 10
+    assert backend_state._replicas[tag].count(
+        version="1", states=[ReplicaState.RUNNING]) == 8
+    assert backend_state._replicas[tag].count(
+        version="1", states=[ReplicaState.STOPPING]) == 2
+    assert not goal_manager.check_complete(goal_2)
+
+    backend_state.update()
+    assert backend_state._replicas[tag].count() == 10
+    assert backend_state._replicas[tag].count(
+        version="1", states=[ReplicaState.RUNNING]) == 8
+    assert backend_state._replicas[tag].count(
+        version="1", states=[ReplicaState.STOPPING]) == 2
+    assert not goal_manager.check_complete(goal_2)
```

Comment:
These two blocks are repeated? if intentional put them in for _ in range(2)

Reply:
oops, accidental